trio/_core/_run.py

```diff
@@ -1092,7 +1100,71 @@ class _RunStatistics:
     run_sync_soon_queue_size = attr.ib()
 
 
-@attr.s(eq=False, hash=False)
+# This holds all the state that gets trampolined back and forth between
+# callbacks when we're running in guest mode.
+#
+# It has to be a separate object from Runner, and Runner *cannot* have hold
+# references to it (directly or indirectly)!
+#
+# The idea is that we want a chance to detect if our host loop quits and stops
+# driving us forward. We detect that by unrolled_run_gen being garbage
+# collected, and hitting its 'except GeneratorExit:' block. So this only
+# happens if unrolled_run_gen is GCed.
+#
+# The Runner state is referenced from the global GLOBAL_RUN_CONTEXT. The only
+# way it gets *un*referenced is by unrolled_run_gen completing, e.g. by being
+# GCed. But if Runner has a direct or indirect reference to it, and the host
+# loop has abandoned it, then this will never happen!
+#
+# So this object can reference Runner, but Runner can't reference it. The only
+# references to it are the "in flight" callback chain on the host loop /
+# worker thread.
+@attr.s(eq=False, hash=False, slots=True)
+class GuestState:
+    runner = attr.ib()
+    run_sync_soon_threadsafe = attr.ib()
+    done_callback = attr.ib()
+    unrolled_run_gen = attr.ib()
+    unrolled_run_next_send = attr.ib(factory=lambda: Value(None))
+
+    def guest_tick(self):
+        locals()[LOCALS_KEY_KI_PROTECTION_ENABLED] = True
+        try:
+            timeout = self.unrolled_run_next_send.send(self.unrolled_run_gen)
+        except StopIteration:
+            # XX if we had KI support, we'd have to do something with it here
+            self.done_callback(self.runner.main_task_outcome)
+            return
+        except TrioInternalError as exc:
+            self.done_callback(Error(exc))
+            return
+
+        # Optimization: try to skip going into the thread if we can avoid it
+        events_outcome = capture(self.runner.io_manager.get_events, 0)
+        if timeout <= 0 or type(events_outcome) is Error or events_outcome.value:
```

Comment:
Got it. I tend to prefer isinstance() because it's understood by type checkers, and it's marginally faster than type-is (although neither one is likely to be noticeable).